crates/ruff_linter/src/rules/flake8_quotes/rules/avoidable_escaped_quote.rs

```diff
@@ -0,0 +1,258 @@
+            }
+            Tok::FStringMiddle {
+                value: string_contents,
+                is_raw,
+            } if !is_raw => {
+                let Some(context) = fstrings.last_mut() else {
+                    continue;
+                };
+                if !context.check_for_escaped_quote {
+                    continue;
+                }
+                if string_contents.contains(quotes_settings.inline_quotes.as_char())
+                    && !string_contents.contains(quotes_settings.inline_quotes.opposite().as_char())
+                {
+                    context.push_fstring_middle_range(tok_range);
+                }
+            }
+            Tok::FStringEnd => {
+                let Some(context) = fstrings.pop() else {
+                    continue;
+                };
+                if context.fstring_middle_ranges.is_empty() {
+                    // There are no `FStringMiddle` tokens containing any escaped
+                    // quotes.
+                    continue;
+                }
+                let mut diagnostic = Diagnostic::new(
+                    AvoidableEscapedQuote,
+                    TextRange::new(context.fstring_start_range.start(), tok_range.end()),
+                );
+                if settings.rules.should_fix(diagnostic.kind.rule()) {
+                    let fstring_start_edit = Edit::range_replacement(
+                        // No need for `r`/`R` as we don't perform the checks
+                        // for raw strings.
+                        format!("f{}", quotes_settings.inline_quotes.opposite().as_char()),
+                        context.fstring_start_range,
+                    );
+                    let fstring_middle_and_end_edits = context
+                        .fstring_middle_ranges
+                        .iter()
+                        .map(|&range| {
+                            Edit::range_replacement(unescaped_string(locator.slice(range)), range)
+                        })
+                        .chain(std::iter::once(
+                            // `FStringEnd` edit
+                            Edit::range_replacement(
+                                quotes_settings
+                                    .inline_quotes
+                                    .opposite()
+                                    .as_char()
+                                    .to_string(),
+                                tok_range,
+                            ),
+                        ));
+                    diagnostic.set_fix(Fix::automatic_edits(
+                        fstring_start_edit,
+                        fstring_middle_and_end_edits,
+                    ));
+                }
+                diagnostics.push(diagnostic);
+            }
+            _ => {}
+        }
+    }
+}
+
+fn unescaped_string(value: &str) -> String {
+    let mut fixed_contents = String::with_capacity(value.len());
+
+    let chars: Vec<char> = value.chars().collect();
```

Comment:
I don't think it's necessary to allocate here. We should be able to implement this with `Peekable`/`Cursor`

Reply:
I didn't look at the implementation in detail but yeah I agree that it can be implemented without allocation. Thanks!

Reply:
Ah ok so this code already exists and was moved.